### PR TITLE
Add support for Ascend NPU

### PR DESCRIPTION
**Pull request by statelesshz**
This PR integrates Ascend NPU hardware capabilities into the sentence-transformers, and enables users to leverage the NPUs for training and inference of embedding models.

Ascend NPU is an AI processor that support AI frameworks like [PyTorch](https://pytorch.org/blog/huawei-joins-pytorch/), TensorFlow, which has already integrated by transformers/accelerate/deepspeed and others popular deepspeed related software and tools.

cc @tomaarsen

### CODE REVIEW

File: sentence_transformers/cross_encoder/CrossEncoder.py [hunk 1]
```diff
@@ -185,10 +184,9 @@ def fit(
         train_dataloader.collate_fn = self.smart_batching_collate
 
         if use_amp:
-            if is_torch_npu_available():
-                scaler = torch.npu.amp.GradScaler()
-            else:
-                scaler = torch.cuda.amp.GradScaler()
```
Comment (tomaarsen) [inline sentence_transformers/cross_encoder/CrossEncoder.py]: I think this if-else is totally fine; it's much simpler than the `device_mod` solution (which also causes a crash if the if-statement is not True).

Could you perhaps revert this change? I think the PR is looking good otherwise.

Comment (statelesshz) [reply sentence_transformers/cross_encoder/CrossEncoder.py]: I've reverted this change. @tomaarsen please take a another look, thx :-)

**Comment by statelesshz** [comment]
I don't know why only ubuntu unit tests failed

**Comment by statelesshz** [comment]
I have tested this patch locally with A100-80G and here are the test results:
```
# python -m pytest -v tests/
============================================================================================ test session starts ============================================================================================
platform linux -- Python 3.8.18, pytest-8.0.0, pluggy-1.4.0 -- /home/miniconda3/envs/w00668292/bin/python
cachedir: .pytest_cache
rootdir: /home/llm/test/sentence-transformers
configfile: pytest.ini
collected 53 items / 12 deselected / 41 selected

tests/test_cmnrl.py::test_cmnrl_same_grad[train_samples_mnrl0-train_samples_cmnrl0-True-1.0-1e-06] PASSED                                                                                             [  2%]
tests/test_cmnrl.py::test_cmnrl_same_grad[train_samples_mnrl1-train_samples_cmnrl1-False-1.0-1e-06] PASSED                                                                                            [  4%]
tests/test_cmnrl.py::test_cmnrl_same_grad[train_samples_mnrl2-train_samples_cmnrl2-True-1000.0-0.001] PASSED                                                                                          [  7%]
tests/test_cmnrl.py::test_rand_context_working[True] PASSED                                                                                                                                           [  9%]
tests/test_cmnrl.py::test_rand_context_working[False] PASSED                                                                                                                                          [ 12%]
tests/test_compute_embeddings.py::test_encode_token_embeddings PASSED                                                                                                                                 [ 14%]
tests/test_compute_embeddings.py::test_encode_single_sentences PASSED                                                                                                                                 [ 17%]
tests/test_compute_embeddings.py::test_encode_normalize PASSED                                                                                                                                        [ 19%]
tests/test_compute_embeddings.py::test_encode_tuple_sentences PASSED                                                                                                                                  [ 21%]
tests/test_cross_encoder.py::test_pretrained_stsb PASSED                                                                                                                                              [ 24%]
tests/test_cross_encoder.py::test_train_stsb PASSED                                                                                                                                                   [ 26%]
tests/test_cross_encoder.py::test_classifier_dropout_is_set PASSED                                                                                                                                    [ 29%]
tests/test_cross_encoder.py::test_classifier_dropout_default_value PASSED                                                                                                                             [ 31%]
tests/test_evaluator.py::test_BinaryClassificationEvaluator_find_best_f1_and_threshold PASSED                                                                                                         [ 34%]
tests/test_evaluator.py::test_BinaryClassificationEvaluator_find_best_accuracy_and_threshold PASSED                                                                                                   [ 36%]
tests/test_evaluator.py::test_LabelAccuracyEvaluator PASSED                                                                                                                                           [ 39%]
tests/test_evaluator.py::test_ParaphraseMiningEvaluator PASSED                                                                                                                                        [ 41%]
tests/test_image_embeddings.py::test_simple_encode PASSED                                                                                                                                             [ 43%]
tests/test_multi_process.py::test_encode_multi_process[False] PASSED                                                                                                                                  [ 46%]
tests/test_multi_process.py::test_encode_multi_process[True] PASSED                                                                                                                                   [ 48%]
tests/test_pretrained_stsb.py::test_bert_base PASSED                                                                                                                                                  [ 51%]
tests/test_pretrained_stsb.py::test_bert_large PASSED                                                                                                                                                 [ 53%]
tests/test_pretrained_stsb.py::test_roberta PASSED                                                                                                                                                    [ 56%]
tests/test_pretrained_stsb.py::test_distilbert PASSED                                                                                                                                                 [ 58%]
tests/test_pretrained_stsb.py::test_multiling PASSED                                                                                                                                                  [ 60%]
tests/test_pretrained_stsb.py::test_mpnet PASSED                                                                                                                                                      [ 63%]
tests/test_pretrained_stsb.py::test_other_models PASSED                                                                                                                                               [ 65%]
tests/test_pretrained_stsb.py::test_msmarco PASSED                                                                                                                                                    [ 68%]
tests/test_pretrained_stsb.py::test_sentence_t5 PASSED                                                                                                                                                [ 70%]
tests/test_sentence_transformer.py::test_load_with_safetensors PASSED                                                                                                                                 [ 73%]
tests/test_sentence_transformer.py::test_to PASSED                                                                                                                                                    [ 75%]
tests/test_sentence_transformer.py::test_save_to_hub PASSED                                                                                                                                           [ 78%]
tests/test_sentence_transformer.py::test_load_with_revision PASSED                                                                                                                                    [ 80%]
tests/test_sentence_transformer.py::test_load_local_without_normalize_directory PASSED                                                                                                                [ 82%]
tests/test_train_stsb.py::test_train_stsb PASSED                                                                                                                                                      [ 85%]
tests/test_train_stsb.py::test_train_nli PASSED                                                                                                                                                       [ 87%]
tests/test_util.py::test_normalize_embeddings PASSED                                                                                                                                                  [ 90%]
tests/test_util.py::test_pytorch_cos_sim PASSED                                                                                                                                                       [ 92%]
tests/test_util.py::test_semantic_search PASSED                                                                                                                                                       [ 95%]
tests/test_util.py::test_paraphrase_mining PASSED                                                                                                                                                     [ 97%]
tests/test_util.py::test_pairwise_scores PASSED                                                                                                                                                       [100%]

============================================================================================= warnings summary ==============================================================================================
tests/test_cmnrl.py: 3 warnings
tests/test_compute_embeddings.py: 4 warnings
tests/test_cross_encoder.py: 4 warnings
tests/test_evaluator.py: 2 warnings
tests/test_image_embeddings.py: 1 warning
tests/test_multi_process.py: 2 warnings
tests/test_pretrained_stsb.py: 23 warnings
tests/test_sentence_transformer.py: 5 warnings
tests/test_train_stsb.py: 2 warnings
tests/test_util.py: 1 warning
  /home/miniconda3/envs/w00668292/lib/python3.8/site-packages/urllib3/connectionpool.py:1099: InsecureRequestWarning: Unverified HTTPS request is being made to host '10.174.181.131'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#tls-warnings
    warnings.warn(

tests/test_compute_embeddings.py::test_encode_token_embeddings
  /home/miniconda3/envs/w00668292/lib/python3.8/site-packages/torch/_utils.py:831: UserWarning: TypedStorage is deprecated. It will be removed in the future and UntypedStorage will be the only storage class. This should only matter to you if you are using storages directly.  To access UntypedStorage directly, use tensor.untyped_storage() instead of tensor.storage()
    return self.fget.__get__(instance, owner)()

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
======================================================================== 41 passed, 12 deselected, 48 warnings in 697.47s (0:11:37) =========================================================================
```

**Comment by tomaarsen** [comment]
Looks clean & complete! Tests pass locally for me too, and thanks for providing test results for your setup too!

- Tom Aarsen
